persistence/pre_persistence/app.go

```diff
@@ -0,0 +1,311 @@
+	"github.com/syndtr/goleveldb/leveldb/util"
+	"google.golang.org/protobuf/proto"
+)
+
+func (m *PrePersistenceContext) GetAppExists(address []byte) (exists bool, err error) {
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	if found := db.Contains(key); !found {
+		return false, nil
+	}
+	bz, err := db.Get(key)
+	if err != nil {
+		return false, err
+	}
+	if bz == nil {
+		return false, nil
+	}
+	if bytes.Contains(bz, DeletedPrefixKey) {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (m *PrePersistenceContext) GetApp(address []byte) (app *App, exists bool, err error) {
+	app = &App{}
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	bz, err := db.Get(key)
+	if err != nil {
+		return nil, false, err
+	}
+	if bz == nil {
+		return nil, false, nil
+	}
+	if bytes.Contains(bz, DeletedPrefixKey) {
+		return nil, false, nil
+	}
+	if err = proto.Unmarshal(bz, app); err != nil {
+		return nil, true, err
+	}
+	return app, true, nil
+}
+
+func (m *PrePersistenceContext) GetAllApps(height int64) (apps []*App, err error) {
+	cdc := Cdc()
+	apps = make([]*App, 0)
+	var it iterator.Iterator
+	if height == m.Height {
+		db := m.Store()
+		it = db.NewIterator(&util.Range{
+			Start: AppPrefixKey,
+			Limit: PrefixEndBytes(AppPrefixKey),
+		})
+	} else {
+		key := HeightKey(height, AppPrefixKey)
+		it = m.Parent.GetCommitDB().NewIterator(&util.Range{
+			Start: key,
+			Limit: PrefixEndBytes(key),
+		})
+	}
+	it.First()
+	defer it.Release()
+	for ; it.Valid(); it.Next() {
+		bz := it.Value()
+		if bytes.Contains(bz, DeletedPrefixKey) {
+			continue
+		}
+		a := App{}
+		if err := cdc.Unmarshal(bz, &a); err != nil {
+			return nil, err
+		}
+		apps = append(apps, &a)
+	}
+	return
+}
+
+func (m *PrePersistenceContext) InsertApplication(address []byte, publicKey []byte, output []byte, paused bool, status int, maxRelays string, stakedTokens string, chains []string, pausedHeight int64, unstakingHeight int64) error {
+	if _, exists, _ := m.GetApp(address); exists {
+		return fmt.Errorf("already exists in world state")
+	}
+	cdc := Cdc()
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	app := App{
+		Address:         address,
+		PublicKey:       publicKey,
+		Paused:          paused,
+		Status:          int32(status),
+		Chains:          chains,
+		MaxRelays:       maxRelays,
+		StakedTokens:    stakedTokens,
+		PausedHeight:    uint64(pausedHeight),
+		UnstakingHeight: unstakingHeight,
+		Output:          output,
+	}
+	bz, err := cdc.Marshal(&app)
+	if err != nil {
+		return err
+	}
+	return db.Put(key, bz)
+}
+
+func (m *PrePersistenceContext) UpdateApplication(address []byte, maxRelaysToAdd string, amountToAdd string, chainsToUpdate []string) error {
+	app, exists, _ := m.GetApp(address)
+	if !exists {
+		return fmt.Errorf("does not exist in world state")
+	}
+	cdc := Cdc()
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	// compute new values
+	stakedTokens, err := StringToBigInt(app.StakedTokens)
```

Comment:
What you're doing here is adding "two strings" which represent big ints.

IMO this warrants a helper that can easily be reused to shorten the code and make it more readable.